fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -174,11 +286,22 @@ void extractKmsUrls(Reference<RESTKmsConnectorCtx> ctx,
 		urlStr.resize(url.GetStringLength());
 		memcpy(urlStr.data(), url.GetString(), url.GetStringLength());
 
-		if (FLOW_KNOBS->REST_LOG_LEVEL >= RESTLogSeverity::INFO) {
-			TraceEvent("RESTExtractDiscoverKmsUrlsAddUrl", ctx->uid).detail("Url", urlStr);
-		}
+		// preserve the KmsUrl stats while (re)discovering KMS URLs, preferable to select the servers with lesser count
+		// of unexpected events in the past
 
-		ctx->kmsUrlHeap.emplace(std::make_shared<KmsUrlCtx>(urlStr));
+		auto itr = urlMap.find(urlStr);
+		if (itr != urlMap.end()) {
+			if (FLOW_KNOBS->REST_LOG_LEVEL >= RESTLogSeverity::INFO) {
+				TraceEvent("RESTDiscoverExistingKmsUrl", ctx->uid).detail("UrlCtx", itr->second.toString());
+			}
+			ctx->kmsUrlStore.kmsUrls.emplace_back(itr->second);
+		} else {
+			auto urlCtx = KmsUrlCtx<KmsUrlPenaltyParams>(urlStr);
+			if (FLOW_KNOBS->REST_LOG_LEVEL >= RESTLogSeverity::INFO) {
+				TraceEvent("RESTDiscoverNewKmsUrl", ctx->uid).detail("UrlCtx", urlCtx.toString());
+			}
+			ctx->kmsUrlStore.kmsUrls.emplace_back(urlCtx);
+		}
```

Comment:
This is a minor optimization but should we do a sort after this? That way if a problematic url is added back first then we don't try to reach out to it before trying the new urls added. Again this case is rare so not sure if its worth modifying the code here